Add configurable outputs to `DiseasyModelOdeSeir` (#207)
ssi-dk/diseasy: vignettes/articles/diseasy-model-ode.Rmd

Reviewer: ```suggestion
- *n_infected*: The "true" number of infected in the data, inferred from observed data and used to initialise the model.
```

Author: In this case, we have a simulation study, so it is really just the true number of infected

Reviewer: ```suggestion
The first step to creating a `?DiseasyModelOdeSeir` model is to define the model scenario.
```

Reviewer: ```suggestion
The scenario specifies the structural assumptions of the model. In this example scenario, as we stated above, we can exactly match the scenario to the data since
```

Reviewer: ```suggestion
the data was created using this same class of models.
```

Reviewer: ```suggestion
In this example scenario we use a simple three age-group stratification of the population in the model.
```

Reviewer: ```suggestion
This particular scenario also includes simple seasonality and waning immunity:
```

Reviewer: ```suggestion
  from $E$ to $I$ and from $I$ to $R$ (i.e. waiting times). The rates determine the expected time spent in each state).
```

Reviewer: ```suggestion
```

Reviewer: ```suggestion
- *overall_infection_risk*: A population-level scaling factor for transmission, controlling the overall risk of infection in the population.
```

The workflow for creating a specific `?DiseasyModelOdeSeir` model would roughly follow these steps:

1) **Defining the scenario.**
  The first step to using most models in `diseasy` is to define the scenario. That is, to configure the functional
  modules that the model templates takes as argument (e.g. `?DiseasyActivity`).
  These steps provide the model template with information about the population, the activity, variants to include etc.
2) **Choosing the model configuration.**
  The model template may have some configurability that you control. In the case of `?DiseasyModelOdeSeir` for example,
  you must choose the number of consecutive $E$, $I$ and $R$ states in the model.
3) **Mapping to observables.**

>   The ODE family of models is in some ways bespoke in the sense that the user needs provide a mapping from the

Reviewer: ```suggestion
  The ODE family of models is in some ways bespoke in the sense that the user needs to provide a mapping from the
```

# Mapping observables to model and back
Being a SEIR model, the `?DiseasyModelOdeSeir` model template runs the dynamics of the

> infections in the population. These dynamics are inferred from the incidence of the

Reviewer: ```suggestion
infections in the population. These dynamics are not usually observed directly. Instead, they must be inferred from one or more observed data sources, such as positive tests or hospital admissions.
```

> disease in the population.

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion

We therefore distinguish between two mappings. First the observed data is used to inter the underlying infection process, and the modelled infection process is then mapped back to the observable quantaties of interest.
```

Mathematically, we map from the set of real-world observables $\vec{O}(t)$ to the

> incidence $I(t)$ via the map $f$:

Reviewer: Måske "N_inf" i stedet for I så man ikke forveksler med "I" state?

> incidence $I(t)$ via the map $f$:

Reviewer: ```suggestion
incidence $N_inf(t)$ via the map $f$:
```

Author: I will change to "Inc(t)"

> $$\vec{O}(t) \overset{f}{\rightarrow} I(t)$$

Reviewer: ```suggestion
$$\vec{O}(t) \overset{f}{\rightarrow} N_inf(t)$$
```

The model then produces a number of observations under the given configuration.

> By default, the model will predict the evolution of the incidence[^1], $\tilde{I}(t)$,

Reviewer: ```suggestion
By default, the model will predict the evolution of the incidence[^1], $\tilde{N_inf}(t)$,
```

but other outputs can be configured (see further below).

[^1]: Measured as the rate of individuals exiting the first `I` state.

To get predictions for other real-world observables, we must map from the incidence
(or other output) to the observable of interest.

> From the model incidence $\tilde{I}(t)$, we then map to the observable estimates

Reviewer: ```suggestion
From the model incidence $\tilde{N_inf}(t)$, we then map to the observable estimates
```

## Mapping from model incidence to observables

> By default, the model can only return predictions for the true number of infected

Reviewer: ```suggestion
By default, the model can only return predictions for the "true" number of infected
```

)
```

### Custom model output
The `diseasy` ODE model templates uses a generic system for generating observables
from the state_vector, $\overline{\psi}$.

> A observable can be configured to either derive from the full state vector

Reviewer: Might be good to add some extra explanation on what can be derived from either two.

$\overline{\psi}$ or the infection matrix $f$ (see the article `SEIR-implementation`
in the online documentation).

In either case, the observable is configured by defining a vector of weights.
For each of such vectors, a "surveillance" state is added to the state_vector which
integrates the inner-product $\overline{\omega}_o \cdot \overline{\psi}$ or
$\overline{\omega}_o \cdot \sum_{rows}\overline{\overline{f}}$ depending on which
element the observable is set to track.

> In the former case, the surveillance states derived from the state_vector where as

Reviewer: ```suggestion
In the former case, the surveillance states are derived from the state_vector (i.e. the configuration of the model) whereas
```

)
```

### Custom model output
The `diseasy` ODE model templates uses a generic system for generating observables
from the state_vector, $\overline{\psi}$.

> A observable can be configured to either derive from either the full state vector

Reviewer: ```suggestion
An observable can be configured to either derive from the full state vector (i.e. the configuration state of the model)
```

are essentially a matrix with weights only for each $I_1$ state[^2] and the observable
tracks the state_vector.

[^2]: The "n_infected" observable does not actually create the additional surveillance states in the implementation
since it can be computed directly from $\overline{\psi}$, but it used here as a illustrative example.

For example, in SEIR case ($N = L = M = 1$) with two age-groups, the weights take the
following form:

$$ \overline{\overline{\Omega}}_{n_{infected}} = \underbrace{\begin{bmatrix}
  0 &  r_i & 0 & 0 & 0 & 0 & 0 & 0 \\
  0 & 0 & 0 & 0 &  r_i & 0 & 0 & 0
\end{bmatrix}}_{\begin{matrix}
  E_1 & I_1 & R_1 & E_2 & I_2 & R_2 & S_1 & S_2
\end{matrix}}
$$
where $r_i$is the rate of transition out of $I$ (i.e. the reciprocal of the time spent
in $I$ state).

As stated above, the model always provides a prediction for the "n_infected" observable
but there are cases where other observables are automatically configured by `diseasy`.

For example, when the model includes a `?DiseasyImmunity` module, observables are
configured to track outcomes related to immunity.

Here, waning target named "hospitalisation" or "death" in the immunity scenario are

> assigned their own observables which is derived from the infection matrix

Reviewer: ```suggestion
assigned their own observables which are derived from the infection matrix
```
